compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs

```diff
@@ -407,11 +408,39 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
 
                 err.span_label(span, cannot_infer_msg);
 
-                let printer = fmt_printer(self, Namespace::TypeNS);
-                let args = printer.comma_sep(generic_args.iter().copied()).unwrap().into_buffer();
+                let args: Result<_, std::fmt::Error> = try {
```

Comment:
Can you instead preprocess `generic_args` first -- replacing args that are not suggestible with a fresh ty or const var? Then pass it to comma_sep.